RFC(0045): Client Side Block Filter Protocol (#375)
nervosnetwork/rfcs: rfcs/0000-client-block-filter/0000-client-block-filter.md

---
Number: "0000"
Category: Standards Track
Status: Proposal
Author: Quake Wang <quake.wang@gmail.com>
Created: 2022-08-23
---

# CKB Client Side Block Filter Protocol

## Abstract

> This RFC describes a light client protocol allowing clients to obtain compact probabilistic filters of block content from full nodes and download full blocks if the filter matches relevant data.

Reviewer: ```suggestion
This RFC describes a block filter protocol that could be used together with [RFC 0044](https://github.com/yangby-cryptape/rfcs/blob/pr/light-client/rfcs/0044-ckb-light-client/0044-ckb-light-client.md). It allows clients to obtain compact probabilistic filters of CKB blocks from full nodes and download full blocks if the filter matches relevant data.
```

## Motivation

> Light clients allow applications to read relevant transactions from the blockchain without incurring the full cost of downloading and validating all data. Such applications seek to simultaneously minimize the trust in peers and the amount of bandwidth, storage space, and computation required. They achieve this by sampling headers through the fly-client protocol, verifying the proofs of work, and following the longest proof-of-work chain. Light clients then download only the blockchain data relevant to them directly from peers and validate inclusion in the header chain. Though clients do not check the validity of all blocks in the longest proof-of-work chain, they rely on miner incentives for security.

Reviewer: This paragraph is describing light client protocol not block filter protocol. Would it be better to contribute it to RFC0044?

Author: This paragraph is about the background of light client, I think it can be kept in this rfc

> Full nodes generate deterministic filters on block data that are served to the client. A light client can then download an entire block if the filter matches the data it is watching for. Since filters are deterministic, they only need to be constructed once and stored on disk, whenever a new block is connected to the chain, this keeps the computation required to serve filters minimal.

Reviewer: ```suggestion
Full nodes generate deterministic filters on block data that are served to the client. A light client can then download an entire block if the filter matches the data it is watching for. Since filters are deterministic, they only need to be constructed once and stored on disk, whenever a new block is appended to the chain. This keeps the computation required to serve filters minimal.
```

    start_number:   Uint64,     // The height of the first block in the requested range
    block_hashes:   Byte32Vec,  // The hashes of the blocks in the range
    filters:        BytesVec,   // The filters of the blocks in the range
}
```

1. The `start_number` SHOULD match the field in the GetBlockFilters request.
2. The `block_hashes` field size should not be larger than 1000.
3. The `block_hashes` and `filters` fields size SHOULD match.

#### GetBlockFilterHashes
`GetBlockFilterHashes` is used to request verifiable filter hashes for a particular range of blocks:

```
struct GetBlockFilterHashes {
    start_number:   Uint64,  // The height of the first block in the requested range
}
```

#### BlockFilterHashes
`BlockFilterHashes` is sent in response to `GetBlockFilterHashes`:

```
table BlockFilterHashes {
    start_number:               Uint64,     // The height of the first block in the requested range
    parent_block_filter_hash:   Byte32,     // The hash of the parent block filter
    block_filter_hashes:        Byte32Vec,  // The hashes of the block filters in the range
}
```

1. The `start_number` SHOULD match the field in the GetBlockFilterHashes request.

> 2. The `block_filter_hashes` field size SHOULD not be larger than 2000.

Reviewer 2: ```suggestion
The `block_filter_hashes` field size SHOULD not exceed 2000
```

```
table BlockFilterCheckPoints {
    start_number:           Uint64,
    block_filter_hashes:    Byte32Vec,
}
```

1. The `start_number` SHOULD match the field in the GetBlockFilterCheckPoints request.
2. The `block_filter_hashes` field size should not be larger than 2000.

### Filter Data Generation

We follow the BIP158 for filter data generation and use the same Golomb-Coded Sets parameters P and M values. The only difference is that we only use cell's lock/type script hash as the filter data:

```
filter.add_element(cell.lock.calc_script_hash().as_slice());
if let Some(type_script) = cell.type_().to_opt() {
    filter.add_element(type_script.calc_script_hash().as_slice());
}
```

### Node Operation

Full nodes MAY opt to support this RFC, such nodes SHOULD treat the filters as an additional index of the blockchain. For each new block that is connected to the main chain, nodes SHOULD generate filters and persist them. Nodes that are missing filters and are already synced with the blockchain SHOULD reindex the chain upon start-up, constructing filters for each block from genesis to the current tip.

> Nodes SHOULD NOT generate filters dynamically on request, as malicious peers may be able to perform DoS attacks by requesting small filters derived from large blocks. This would require an asymmetical amount of I/O on the node to compute and serve.

Reviewer 2: ```suggestion
Nodes SHOULD NOT generate filters dynamically on request, as malicious peers may be able to perform DoS attacks by requesting small filters derived from large blocks. This would require an asymmetrical amount of I/O on the node to compute and serve.
```

Nodes MAY prune block data after generating and storing all filters for a block.

### Client Operation

This section provides recommendations for light clients to download filters with maximal security.

> Clients SHOULD first sync with the full nodes by verifiying the best chain tip through the fly-client protocol before downloading any filters or filter hashes. Clients SHOULD disconnect any outbound peers whose best chain has significantly less work than the known longest chain.

Reviewer 2: ```suggestion
Clients SHOULD first sync with the full nodes by verifying the best chain tip through the fly-client protocol before downloading any filters or filter hashes. Clients SHOULD disconnect any outbound peers whose best chain has significantly less work than the known longest chain.

```

---

> Number: "0000"

Author: ```suggestion
Number: "0045"
```

> Number: "0045"

Reviewer 3: File path still uses 0000
